src/ContainerTrait.php

```diff
@@ -14,22 +39,152 @@ public function _unique_element($desired = null)
+
+            // ask element on his preferred name, then make it unique.
+            $cn = $element->getDesiredName();
+            $args[0] = $this->_unique_element($cn);
+        } else {
+
+            // generate name based on the class
+            $cn = str_replace('\\', '_', strtolower(get_class($element)));
+            $args[0] = $this->_unique_element($cn);
+        }
+
+        // Maybe element already exists
+        if (isset($this->elements[$args[0]])) {
+            throw new Exception([
+                'Element with requested name already exists',
+                'element'=>$element,
+                'name'=>$args[0],
+                'this'=>$this,
+                'arg2'=>$args
+            ]);
+        }
+
+        if(isset($this->_appScope) && isset($element->_appScope)) {
+            $element->app = $this->app;
+        }
+
+        $element->owner = $this;
+        $element->short_name = $args[0];
+
+
+        if(isset($element->_trackableTrait)) {
+
+            $element->name = $this->_shorten($this->name.'_'.$element->short_name);
+
+            $this->elements[$element->short_name] = $element;
+        } else {
+            // dont store extra reference to models and controlers
+            // for purposes of better garbage collection.
+            $this->elements[$element->short_name] = true;
+        }
+        
+        return $element;
     }
 
     /**
-     * Extension to add() method which will perform container tracking
+     * Remove child element if it exists.
+     *
+     * @param string $short_name short name of the element
+     *
+     * @return $this
      */
-    protected function _add_Container() {
-    }
+    public function removeElement($short_name)
+    {
 
-    protected function _shorten () {
+        if (is_object($short_name)) {
+            $short_name = $short_name->short_name;
+        }
+        unset($this->elements[$short_name]);
+        return $this;
     }
 
-    function getElement() {
+    /**
+     * Method used internally for shortening object names.
+     *
+     * @param string $desired Desired name of new object.
+     *
+     * @return string Shortened name of new object.
+     */
+    protected function _shorten ($desired) {
+        if (
+            isset($this->_appScope) &&
+            $this->app->max_name_length &&
+            strlen($desired) > $this->app->max_name_length
+        ) {
+            // $len is the amount to chomp. It must divide
+            // by max_name_length, so that we keep chomping
+            // by consistent amounts
+            $len = strlen($desired);
+            $len -= ($len % $this->app->max_name_length);
+
+            $key = substr($desired, 0, $len);
+            $rest = substr($desired, $len);
+
+            if (!isset($this->app->unique_hashes[$key])) {
+                $this->app->unique_hashes[$key] = dechex(crc32($key));
+            }
+            $desired = $this->app->unique_hashes[$key].'__'.$rest;
```

Comment:
I have refactored the logic of the method. It now uses chunks of max_name_length to cut. Assume max-name_length = 10. And length of the string is 34.

The $len will be 30 here, so it will cut off 30 and convert that into the key. The rest 4 characters will remain. 

True that after re-attaching the hash on the front it will be more than 10, if we increase len, we should check to make sure we don't leave empty space at the end. 

This needs more test-cases.

The reason why i'm evening the cuts is to avoid situation like this:

 - 281937_bla_hello_world
 - 199402_la_hello_world2

This works with the current ATK cutting too, but for situations when short_name is much longer - it won't cut off enough.

 - [ ] Add more test-scripts an change this method if necessary.